### PR TITLE
feat(compact): allow batch fetch_entry

### PR DESCRIPTION
**Pull request by ackRow**
Work in progress on the issue https://github.com/Cosmian/findex/issues/9

### CODE REVIEW

File: src/core/compact.rs [hunk 1]
```diff
@@ -64,6 +67,7 @@ pub trait FindexCompact<
         master_key: &KeyingMaterial<MASTER_KEY_LENGTH>,
         new_master_key: &KeyingMaterial<MASTER_KEY_LENGTH>,
         label: &Label,
+        fetch_entry_batch_size: usize,
```
Comment (tbrezot) [inline src/core/compact.rs:70]: missing documentation

File: src/core/compact.rs [hunk 2]
```diff
@@ -85,153 +89,172 @@ pub trait FindexCompact<
...
-        // `chain_table_uids_to_remove`) and send them to the callback to update the
-        // database.
-        let mut chain_table_adds = EncryptedTable::default();
-        let mut entry_table = EntryTable::decrypt::<BLOCK_LENGTH, DEM_KEY_LENGTH, DemScheme>(
-            &k_value,
-            &encrypted_entry_table,
-        )?;
-        // Entry Table items indexing empty chains should be removed.
-        let mut entry_table_uids_to_drop = Vec::new();
+        let all_uids_vec = Vec::from_iter(all_uids.clone());
```
Comment (tbrezot) [inline src/core/compact.rs:92]: You don't need to keep the old version, this should save a copy.

File: src/core/compact.rs [hunk 3]
```diff
@@ -85,153 +89,172 @@ pub trait FindexCompact<
...
         .ceil() as usize;
-        let entry_table_items_to_reindex = entry_table
-            .keys()
+        let entry_table_uids_to_reindex = all_uids
+            .iter()
             .choose_multiple(&mut rng, n_lines)
             .into_iter()
             .cloned()
-            .collect::<Vec<_>>();
+            .collect::<HashSet<_>>();
```
Comment (tbrezot) [inline src/core/compact.rs:107]: You shouldn't need a `HashSet` here

File: src/core/compact.rs [hunk 4]
```diff
@@ -85,153 +89,172 @@ pub trait FindexCompact<
...
-        //
-        // Remove all reindexed Chain Table items. Chains are recomputed entirely.
-        //
-        let chain_table_uids_to_remove = chains_to_reindex
-            .values()
-            .flat_map(|chain| chain.iter().map(|(k, _)| k))
-            .cloned()
-            .collect::<HashSet<_>>();
+            let encrypted_entry_table = self.fetch_entry_table(&slice_uids).await?;
+            println!("table: {encrypted_entry_table:?}");
```
Comment (tbrezot) [inline src/core/compact.rs:118]: Could `compact()` look like that?
```
let all_uids = fetch_entry_table_uids();
for uid_batch in all_uids.by_batch() {
    let entry_table_batch = fetch_entry_table(uid_batch).await?;
    compact_subroutine(entry_table_batch)
}
```

*Note*: this is pseudo-code, it seems there is no function such as `.by_batch()` in the stable version, or at least I couldn't find one.

File: src/interfaces/pyo3/py_api.rs [hunk 5]
```diff
@@ -486,6 +493,7 @@ impl InternalFindex {
             &master_key.0,
             &new_master_key.0,
             &new_label.0,
+            2,
```
Comment (tbrezot) [inline src/interfaces/pyo3/py_api.rs:496]: This is a draft but you should remember to change that.

File: src/interfaces/tests.rs [hunk 6]
```diff
@@ -506,7 +509,7 @@ async fn test_findex() -> Result<(), FindexErr> {
         new_label = Label::random(&mut rng);
         let new_master_key = KeyingMaterial::new(&mut rng);
         findex
-            .compact(i, &master_key, &new_master_key, &new_label)
+            .compact(i, &master_key, &new_master_key, &new_label, 2)
```
Comment (tbrezot) [inline src/interfaces/tests.rs:512]: Use a constant instead. Same everywhere.

File: src/core/compact.rs (hunk 3 above)
Comment (ackRow) [reply src/core/compact.rs:109]: The hashset is used for computing an intersection

**Pull request closed without merging**